dattri/algorithm/trak.py

```diff
@@ -127,14 +133,22 @@ def cache(
                 desc="calculating gradient of training set...",
                 leave=False,
             ):
-                train_batch_data = tuple(data.to(self.device) for data in train_data)
+                # TODO: reorganize the data pre-grad processing.
+                if isinstance(train_data, (tuple, list)):
+                    train_batch_data = tuple(
+                        data.to(self.device) for data in train_data
+                    )
+                else:
```

Comment:
per your description, so if `train_data` is not a tuple of a list, then it must be a dict?
And why if `train_data` is a dict, then we don't need to put data onto `self.device`?
In addition, do we need to restrict the type of `train_data` here?

Reply:
I leave this to future PR. As I stated in TODO (`TODO: reorganize the data pre-grad processing.`). I will raise a PR for future developers to do this. I understand now the preprocessing is very weak but it seems too much for this PR.

Reply:
Make sense